<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.4"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>HogProf: pyoma.browser.db.Database Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">HogProf<span id="projectnumber">&#160;0.1</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.4 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>pyoma</b></li><li class="navelem"><b>browser</b></li><li class="navelem"><b>db</b></li><li class="navelem"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html">Database</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classpyoma_1_1browser_1_1db_1_1Database-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">pyoma.browser.db.Database Class Reference</div></div>
</div><!--header-->
<div class="contents">
<div class="dynheader">
Inheritance diagram for pyoma.browser.db.Database:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classpyoma_1_1browser_1_1db_1_1Database.png" alt=""/>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a156a953e58f0072222d82eafed8df3fe"><td class="memItemLeft" align="right" valign="top"><a id="a156a953e58f0072222d82eafed8df3fe" name="a156a953e58f0072222d82eafed8df3fe"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>__init__</b> (self, db)</td></tr>
<tr class="separator:a156a953e58f0072222d82eafed8df3fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a287aba2f5654828223c15de4e4464456"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a287aba2f5654828223c15de4e4464456">gene_ontology</a> (self)</td></tr>
<tr class="separator:a287aba2f5654828223c15de4e4464456"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7709c6ca40041c9a858153cb28788e1c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a7709c6ca40041c9a858153cb28788e1c">load_gene_ontology</a> (self, factory=None, rels=None)</td></tr>
<tr class="separator:a7709c6ca40041c9a858153cb28788e1c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abac2267fbad28c4fa6b9182d66141e36"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#abac2267fbad28c4fa6b9182d66141e36">get_hdf5_handle</a> (self)</td></tr>
<tr class="separator:abac2267fbad28c4fa6b9182d66141e36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad695adf3059ce1e672b9b09c4983074b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#ad695adf3059ce1e672b9b09c4983074b">get_conversion_date</a> (self)</td></tr>
<tr class="separator:ad695adf3059ce1e672b9b09c4983074b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae069aa87220ee4c03768cdd904103875"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#ae069aa87220ee4c03768cdd904103875">ensure_entry</a> (self, entry)</td></tr>
<tr class="separator:ae069aa87220ee4c03768cdd904103875"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad0196cd3a387f06fdbde097e32a1d4c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#aad0196cd3a387f06fdbde097e32a1d4c">entry_by_entry_nr</a> (self, entry_nr)</td></tr>
<tr class="separator:aad0196cd3a387f06fdbde097e32a1d4c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2e9290c12b957be9f184694fa2e49f48"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a2e9290c12b957be9f184694fa2e49f48">all_proteins_of_genome</a> (self, genome)</td></tr>
<tr class="separator:a2e9290c12b957be9f184694fa2e49f48"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a17e6f23f51a65c802055778f8aa3abdf"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a17e6f23f51a65c802055778f8aa3abdf">main_isoforms</a> (self, genome)</td></tr>
<tr class="separator:a17e6f23f51a65c802055778f8aa3abdf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76b1c0205209542bebf232e5ec143587"><td class="memItemLeft" align="right" valign="top"><a id="a76b1c0205209542bebf232e5ec143587" name="a76b1c0205209542bebf232e5ec143587"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_splicing_variants</b> (self, entry)</td></tr>
<tr class="separator:a76b1c0205209542bebf232e5ec143587"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab65792120491bf66f65a4ec543f4a066"><td class="memItemLeft" align="right" valign="top"><a id="ab65792120491bf66f65a4ec543f4a066" name="ab65792120491bf66f65a4ec543f4a066"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>count_vpairs</b> (self, entry_nr)</td></tr>
<tr class="separator:ab65792120491bf66f65a4ec543f4a066"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab4632b7336e3582189dfa6bf1bae434e"><td class="memItemLeft" align="right" valign="top"><a id="ab4632b7336e3582189dfa6bf1bae434e" name="ab4632b7336e3582189dfa6bf1bae434e"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>count_homoeologs</b> (self, entry_nr)</td></tr>
<tr class="separator:ab4632b7336e3582189dfa6bf1bae434e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afa759c54091757ae9999390e074dc793"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#afa759c54091757ae9999390e074dc793">get_vpairs</a> (self, entry_nr)</td></tr>
<tr class="separator:afa759c54091757ae9999390e074dc793"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a098a5ed5cf1b8eff7cb64a11395d708b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a098a5ed5cf1b8eff7cb64a11395d708b">get_within_species_paralogs</a> (self, entry_nr)</td></tr>
<tr class="separator:a098a5ed5cf1b8eff7cb64a11395d708b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6d4e108af8cb8d0417a1c081d446b9dd"><td class="memItemLeft" align="right" valign="top"><a id="a6d4e108af8cb8d0417a1c081d446b9dd" name="a6d4e108af8cb8d0417a1c081d446b9dd"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_homoeologs</b> (self, entry_nr)</td></tr>
<tr class="separator:a6d4e108af8cb8d0417a1c081d446b9dd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac2688c4b6c269f99fc8974fb97ac47ab"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#ac2688c4b6c269f99fc8974fb97ac47ab">neighbour_genes</a> (self, entry_nr, window=1)</td></tr>
<tr class="separator:ac2688c4b6c269f99fc8974fb97ac47ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af13515258c6cc46aa181ffa077f2d521"><td class="memItemLeft" align="right" valign="top"><a id="af13515258c6cc46aa181ffa077f2d521" name="af13515258c6cc46aa181ffa077f2d521"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>parse_hog_id</b> (self, hog_id)</td></tr>
<tr class="separator:af13515258c6cc46aa181ffa077f2d521"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac107572cddc042a6726f88367b5defbd"><td class="memItemLeft" align="right" valign="top"><a id="ac107572cddc042a6726f88367b5defbd" name="ac107572cddc042a6726f88367b5defbd"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>hog_family</b> (self, entry)</td></tr>
<tr class="separator:ac107572cddc042a6726f88367b5defbd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa214c4a5db019abcf849cc3dc469199c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#aa214c4a5db019abcf849cc3dc469199c">hog_levels_of_fam</a> (self, fam_nr)</td></tr>
<tr class="separator:aa214c4a5db019abcf849cc3dc469199c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a49b1d289a42046eef3369f0f92bfbe0a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a49b1d289a42046eef3369f0f92bfbe0a">get_subhogids_at_level</a> (self, fam_nr, level)</td></tr>
<tr class="separator:a49b1d289a42046eef3369f0f92bfbe0a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a191627558268ec8561af49ae1f1ff975"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a191627558268ec8561af49ae1f1ff975">member_of_hog_id</a> (self, hog_id, level=None)</td></tr>
<tr class="separator:a191627558268ec8561af49ae1f1ff975"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af245ce87a9323b696b56ef3ca8d69811"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#af245ce87a9323b696b56ef3ca8d69811">iter_members_of_hog_id</a> (self, hog_id)</td></tr>
<tr class="separator:af245ce87a9323b696b56ef3ca8d69811"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abafa5fc4a4b8d2f8df8f787d5119deca"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#abafa5fc4a4b8d2f8df8f787d5119deca">member_of_fam</a> (self, fam)</td></tr>
<tr class="separator:abafa5fc4a4b8d2f8df8f787d5119deca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b6d244495e0d8f8d5bd2097d87101d1"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a8b6d244495e0d8f8d5bd2097d87101d1">hog_members</a> (self, entry, level)</td></tr>
<tr class="separator:a8b6d244495e0d8f8d5bd2097d87101d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9ee96b8524c16ef3b914480eba6e8ffb"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a9ee96b8524c16ef3b914480eba6e8ffb">hog_members_from_hog_id</a> (self, hog_id, level)</td></tr>
<tr class="separator:a9ee96b8524c16ef3b914480eba6e8ffb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae3133157ae6f2c74418acee3b801d3b0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#ae3133157ae6f2c74418acee3b801d3b0">get_orthoxml</a> (self, fam)</td></tr>
<tr class="separator:ae3133157ae6f2c74418acee3b801d3b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a40877a4e516d4d1fdce2e1582e658b2e"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a40877a4e516d4d1fdce2e1582e658b2e">oma_group_members</a> (self, group_id)</td></tr>
<tr class="separator:a40877a4e516d4d1fdce2e1582e658b2e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54d6c109ca4bc04ff23fd704b9edb97d"><td class="memItemLeft" align="right" valign="top"><a id="a54d6c109ca4bc04ff23fd704b9edb97d" name="a54d6c109ca4bc04ff23fd704b9edb97d"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>resolve_oma_group</b> (self, group_id)</td></tr>
<tr class="separator:a54d6c109ca4bc04ff23fd704b9edb97d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4867adaf5cca9f18371f1b985097cc22"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a4867adaf5cca9f18371f1b985097cc22">oma_group_metadata</a> (self, group_nr)</td></tr>
<tr class="separator:a4867adaf5cca9f18371f1b985097cc22"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8d59407836e4efb2a2ec2bdb5d0201e"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#ae8d59407836e4efb2a2ec2bdb5d0201e">get_nr_oma_groups</a> (self)</td></tr>
<tr class="separator:ae8d59407836e4efb2a2ec2bdb5d0201e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa1d218f898e16fd21d120a73c89a2f68"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#aa1d218f898e16fd21d120a73c89a2f68">get_nr_toplevel_hogs</a> (self)</td></tr>
<tr class="separator:aa1d218f898e16fd21d120a73c89a2f68"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acd31bc82756a5510d212946f2d9ad3ac"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#acd31bc82756a5510d212946f2d9ad3ac">group_size_histogram</a> (self, typ=None)</td></tr>
<tr class="separator:acd31bc82756a5510d212946f2d9ad3ac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a509bd1c4614c588429afc924c5a6fddc"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a509bd1c4614c588429afc924c5a6fddc">get_sequence</a> (self, entry)</td></tr>
<tr class="separator:a509bd1c4614c588429afc924c5a6fddc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a89d544a9bce504314ffd31986f08f62c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a89d544a9bce504314ffd31986f08f62c">get_cdna</a> (self, entry)</td></tr>
<tr class="separator:a89d544a9bce504314ffd31986f08f62c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af01129ff322fbf7363f99b75c963281f"><td class="memItemLeft" align="right" valign="top"><a id="af01129ff322fbf7363f99b75c963281f" name="af01129ff322fbf7363f99b75c963281f"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_description</b> (self, entry)</td></tr>
<tr class="separator:af01129ff322fbf7363f99b75c963281f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e3651e1afe5d8a80280f271ab056f46"><td class="memItemLeft" align="right" valign="top"><a id="a5e3651e1afe5d8a80280f271ab056f46" name="a5e3651e1afe5d8a80280f271ab056f46"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_release_name</b> (self)</td></tr>
<tr class="separator:a5e3651e1afe5d8a80280f271ab056f46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1abdf43aa28f078a0e1b5322096c97f1"><td class="memItemLeft" align="right" valign="top"><a id="a1abdf43aa28f078a0e1b5322096c97f1" name="a1abdf43aa28f078a0e1b5322096c97f1"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_exons</b> (self, entry_nr)</td></tr>
<tr class="separator:a1abdf43aa28f078a0e1b5322096c97f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8d06ab8bf9dd62583ceaa55a4ffc863c"><td class="memItemLeft" align="right" valign="top"><a id="a8d06ab8bf9dd62583ceaa55a4ffc863c" name="a8d06ab8bf9dd62583ceaa55a4ffc863c"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_domains</b> (self, entry_nr)</td></tr>
<tr class="separator:a8d06ab8bf9dd62583ceaa55a4ffc863c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44410adcb837deb20a77bdd90c3e6251"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a44410adcb837deb20a77bdd90c3e6251">get_representative_entry_of_hog</a> (self, fam)</td></tr>
<tr class="separator:a44410adcb837deb20a77bdd90c3e6251"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acc06abc8c06989bfd9f65ef2f55602f1"><td class="memItemLeft" align="right" valign="top"><a id="acc06abc8c06989bfd9f65ef2f55602f1" name="acc06abc8c06989bfd9f65ef2f55602f1"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_prevalent_domains</b> (self, fam)</td></tr>
<tr class="separator:acc06abc8c06989bfd9f65ef2f55602f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5de3407e15d9d0f435439c06d4ecf414"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpyoma_1_1browser_1_1db_1_1Database.html#a5de3407e15d9d0f435439c06d4ecf414">get_gene_ontology_annotations</a> (self, entry_nr, stop=None, as_dataframe=False, as_gaf=False)</td></tr>
<tr class="separator:a5de3407e15d9d0f435439c06d4ecf414"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-attribs" name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:af470f41e286ca68193b31fbc404640cd"><td class="memItemLeft" align="right" valign="top"><a id="af470f41e286ca68193b31fbc404640cd" name="af470f41e286ca68193b31fbc404640cd"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>db</b></td></tr>
<tr class="separator:af470f41e286ca68193b31fbc404640cd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5949655957a346fc97f4079d90c9c928"><td class="memItemLeft" align="right" valign="top"><a id="a5949655957a346fc97f4079d90c9c928" name="a5949655957a346fc97f4079d90c9c928"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>db_schema_version</b></td></tr>
<tr class="separator:a5949655957a346fc97f4079d90c9c928"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a70e7b0c9a74b337fd94f24b6214b31b5"><td class="memItemLeft" align="right" valign="top"><a id="a70e7b0c9a74b337fd94f24b6214b31b5" name="a70e7b0c9a74b337fd94f24b6214b31b5"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>seq_search</b></td></tr>
<tr class="separator:a70e7b0c9a74b337fd94f24b6214b31b5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a57fd9d1269f3e77f79572a0442a95441"><td class="memItemLeft" align="right" valign="top"><a id="a57fd9d1269f3e77f79572a0442a95441" name="a57fd9d1269f3e77f79572a0442a95441"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>id_resolver</b></td></tr>
<tr class="separator:a57fd9d1269f3e77f79572a0442a95441"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae011b537a9aecc53ee0dbd0e5d7c673d"><td class="memItemLeft" align="right" valign="top"><a id="ae011b537a9aecc53ee0dbd0e5d7c673d" name="ae011b537a9aecc53ee0dbd0e5d7c673d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>id_mapper</b></td></tr>
<tr class="separator:ae011b537a9aecc53ee0dbd0e5d7c673d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4291e9117cea19452155bacedc35b842"><td class="memItemLeft" align="right" valign="top"><a id="a4291e9117cea19452155bacedc35b842" name="a4291e9117cea19452155bacedc35b842"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>tax</b></td></tr>
<tr class="separator:a4291e9117cea19452155bacedc35b842"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a456a6f7b841c2b22e3bc1e11a1817ca5"><td class="memItemLeft" align="right" valign="top"><a id="a456a6f7b841c2b22e3bc1e11a1817ca5" name="a456a6f7b841c2b22e3bc1e11a1817ca5"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>format_hogid</b></td></tr>
<tr class="separator:a456a6f7b841c2b22e3bc1e11a1817ca5"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-static-attribs" name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a288083c4275e6be71e9562cbad7d9fa2"><td class="memItemLeft" align="right" valign="top"><a id="a288083c4275e6be71e9562cbad7d9fa2" name="a288083c4275e6be71e9562cbad7d9fa2"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>EXPECTED_DB_SCHEMA</b> = &quot;3.2&quot;</td></tr>
<tr class="separator:a288083c4275e6be71e9562cbad7d9fa2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><pre class="fragment">This is the main interface to the oma database. Queries
will typically be issued by methods of this object. Typically
the result of queries will be :py:class:`numpy.recarray` objects.</pre> </div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a2e9290c12b957be9f184694fa2e49f48" name="a2e9290c12b957be9f184694fa2e49f48"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2e9290c12b957be9f184694fa2e49f48">&#9670;&nbsp;</a></span>all_proteins_of_genome()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.all_proteins_of_genome </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>genome</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">return all protein entries of a genome</pre> 
</div>
</div>
<a id="ae069aa87220ee4c03768cdd904103875" name="ae069aa87220ee4c03768cdd904103875"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae069aa87220ee4c03768cdd904103875">&#9670;&nbsp;</a></span>ensure_entry()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.ensure_entry </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">This method allows to use an entry or an entry_nr.

If necessary it will load the entry from the entry_nr,
otherwise returning the same object again.

:param entry: the entry_nr of a protein to be loaded or a
    protein entry.</pre> 
</div>
</div>
<a id="aad0196cd3a387f06fdbde097e32a1d4c" name="aad0196cd3a387f06fdbde097e32a1d4c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aad0196cd3a387f06fdbde097e32a1d4c">&#9670;&nbsp;</a></span>entry_by_entry_nr()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.entry_by_entry_nr </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry_nr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Returns the entry from the /Protein/Entries table
corresponding to entry_nr.

:param int entry_nr: a numeric identifier for the protein
    entry</pre> 
</div>
</div>
<a id="a287aba2f5654828223c15de4e4464456" name="a287aba2f5654828223c15de4e4464456"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a287aba2f5654828223c15de4e4464456">&#9670;&nbsp;</a></span>gene_ontology()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.gene_ontology </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns GeneOntology object containing hierarchy
of terms using the is_a and part_of relations. See
:meth:`load_gene_ontology` to parametrize the
creation of GeneOntology object.</pre> 
</div>
</div>
<a id="a89d544a9bce504314ffd31986f08f62c" name="a89d544a9bce504314ffd31986f08f62c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a89d544a9bce504314ffd31986f08f62c">&#9670;&nbsp;</a></span>get_cdna()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_cdna </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get the protein sequence of a given entry as a string</pre> 
</div>
</div>
<a id="ad695adf3059ce1e672b9b09c4983074b" name="ad695adf3059ce1e672b9b09c4983074b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad695adf3059ce1e672b9b09c4983074b">&#9670;&nbsp;</a></span>get_conversion_date()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_conversion_date </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">return the conversion end date from the DB attributes</pre> 
</div>
</div>
<a id="a5de3407e15d9d0f435439c06d4ecf414" name="a5de3407e15d9d0f435439c06d4ecf414"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5de3407e15d9d0f435439c06d4ecf414">&#9670;&nbsp;</a></span>get_gene_ontology_annotations()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_gene_ontology_annotations </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry_nr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>stop</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>as_dataframe</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>as_gaf</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Retrieve the gene ontology annotations for an entry or entry_range

The method returns the gene ontology annotations stored in the database
for a given entry (if `stop` parameter is not provided) or for all the
entries between [entry_nr, stop). Like in slices, the stop entry_nr is
not inclusive, where as the entry_nr - the start of the slice - is.

By default the result are returned as numpy arrays of type
:class:`tablefmt.GeneOntologyTable`. If as_dataframe is set to true, the
result will be a pandas dataframe, and if as_gaf is set to true, a gaf
formatted text file with the annotations is returned.

:param int entry_nr: numeric protein entry
</pre> 
</div>
</div>
<a id="abac2267fbad28c4fa6b9182d66141e36" name="abac2267fbad28c4fa6b9182d66141e36"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abac2267fbad28c4fa6b9182d66141e36">&#9670;&nbsp;</a></span>get_hdf5_handle()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_hdf5_handle </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">return the handle to the database hdf5 file</pre> 
</div>
</div>
<a id="ae8d59407836e4efb2a2ec2bdb5d0201e" name="ae8d59407836e4efb2a2ec2bdb5d0201e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae8d59407836e4efb2a2ec2bdb5d0201e">&#9670;&nbsp;</a></span>get_nr_oma_groups()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_nr_oma_groups </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns the number of OMA Groups in the database</pre> 
</div>
</div>
<a id="aa1d218f898e16fd21d120a73c89a2f68" name="aa1d218f898e16fd21d120a73c89a2f68"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa1d218f898e16fd21d120a73c89a2f68">&#9670;&nbsp;</a></span>get_nr_toplevel_hogs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_nr_toplevel_hogs </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns the number of toplevel hogs, i.e. roothogs</pre> 
</div>
</div>
<a id="ae3133157ae6f2c74418acee3b801d3b0" name="ae3133157ae6f2c74418acee3b801d3b0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae3133157ae6f2c74418acee3b801d3b0">&#9670;&nbsp;</a></span>get_orthoxml()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_orthoxml </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>fam</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns the orthoxml of a given toplevel HOG family

:param fam: numeric id of requested toplevel hog</pre> 
</div>
</div>
<a id="a44410adcb837deb20a77bdd90c3e6251" name="a44410adcb837deb20a77bdd90c3e6251"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a44410adcb837deb20a77bdd90c3e6251">&#9670;&nbsp;</a></span>get_representative_entry_of_hog()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_representative_entry_of_hog </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>fam</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Get the information of the representative entry for a given family (roothog).

For each family we select a represenative entry that has the most prevalent
domain architecture. This method returns the entry_nr that we selected, together
with the domain architecture and its prevalence. In case no representative entry
has been found, the method raises an :class:`NoReprEntry` Exception.

:param int fam: The numeric family number.</pre> 
</div>
</div>
<a id="a509bd1c4614c588429afc924c5a6fddc" name="a509bd1c4614c588429afc924c5a6fddc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a509bd1c4614c588429afc924c5a6fddc">&#9670;&nbsp;</a></span>get_sequence()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_sequence </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get the protein sequence of a given entry as a string

:param entry: the entry or entry_nr for which the sequence is requested</pre> 
</div>
</div>
<a id="a49b1d289a42046eef3369f0f92bfbe0a" name="a49b1d289a42046eef3369f0f92bfbe0a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a49b1d289a42046eef3369f0f92bfbe0a">&#9670;&nbsp;</a></span>get_subhogids_at_level()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_subhogids_at_level </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>fam_nr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>level</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get all the hog ids within a given family at a given taxonomic
level of interest.

After a duplication in an ancestor lineage, there exists multiple
sub-hogs for any taxonomic level after the duplication. This method
allows to get the list of hogids at the requested taxonomic level.

E.g. assume in family 1 (HOG:0000001) there has been a duplication
between Eukaryota and Metazoa. this method would return for
get_subhogids_at_level(1, 'Eukaryota') --&gt; ['HOG:0000001']
and for
get_subhogids_at_level(1, 'Metazoa') --&gt; ['HOG:0000001.1a', 'HOG:0000001.1b']

:param fam_nr: the numeric family id
:param level: the taxonomic level of interest</pre> 
</div>
</div>
<a id="afa759c54091757ae9999390e074dc793" name="afa759c54091757ae9999390e074dc793"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afa759c54091757ae9999390e074dc793">&#9670;&nbsp;</a></span>get_vpairs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_vpairs </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry_nr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns the verified pairs of a query protein.

This method returns an instance of a :class:`numpy.recarray` class
containing the verified pairs of a query protein entry.
The returned array contains columns with EntryNr1 and EntryNr2 to
identify the pair together with RelType (indicating the subtype of
orthology), the alignment score and the distance. The score and
distance will be set to -1 if unknown.

:param int entry_nr: the numeric entry_nr of the query protein.</pre> 
</div>
</div>
<a id="a098a5ed5cf1b8eff7cb64a11395d708b" name="a098a5ed5cf1b8eff7cb64a11395d708b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a098a5ed5cf1b8eff7cb64a11395d708b">&#9670;&nbsp;</a></span>get_within_species_paralogs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.get_within_species_paralogs </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry_nr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns the within species paralogs of a given entry

This method returns a :class:`numpy.recarray` instance
containing the close paralogs. Close paralogs are within
species paralogs that are inparalogs to at least one
ortholog of the query gene in OMA.

The returned array contains columns with EntryNr1 and EntryNr2 to
identify the pair together with RelType (indicating the subtype of
paralogy), the alignment score and the distance. The score and
distance will be set to -1 if unknown.

:param int entry_nr: the numeric entry_id of the query protein</pre> 
</div>
</div>
<a id="acd31bc82756a5510d212946f2d9ad3ac" name="acd31bc82756a5510d212946f2d9ad3ac"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acd31bc82756a5510d212946f2d9ad3ac">&#9670;&nbsp;</a></span>group_size_histogram()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.group_size_histogram </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>typ</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns a table with two columns, e.g. Size and Count.

if typ is set to 'oma' or not set, then the data for the
oma groups is returned. if it is set to 'hog', the data for
the rootlevel hogs is returned.

:param typ: either 'oma' or 'hog', defaults to 'oma'</pre> 
</div>
</div>
<a id="aa214c4a5db019abcf849cc3dc469199c" name="aa214c4a5db019abcf849cc3dc469199c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa214c4a5db019abcf849cc3dc469199c">&#9670;&nbsp;</a></span>hog_levels_of_fam()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.hog_levels_of_fam </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>fam_nr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get all taxonomic levels covered by a family.

The family coresponds to the toplevel numeric id of a HOG,
i.e. for HOG:002421 the fam_nr should be 2421. If a HOG
covers a certain level more than once, it will be returned
several times.

:param fam_nr: the numeric id of the family (== Toplevel HOG)
</pre> 
</div>
</div>
<a id="a8b6d244495e0d8f8d5bd2097d87101d1" name="a8b6d244495e0d8f8d5bd2097d87101d1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8b6d244495e0d8f8d5bd2097d87101d1">&#9670;&nbsp;</a></span>hog_members()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.hog_members </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>level</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get hog members with respect to a given taxonomic level.

The method will return a list of protein entries that are all
member of the same hog with respect to the taxonomic range
of interest.

:param entry: an entry or entry_nr of a query protein
:param level: the taxonomic level of interest</pre> 
</div>
</div>
<a id="a9ee96b8524c16ef3b914480eba6e8ffb" name="a9ee96b8524c16ef3b914480eba6e8ffb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9ee96b8524c16ef3b914480eba6e8ffb">&#9670;&nbsp;</a></span>hog_members_from_hog_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.hog_members_from_hog_id </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>hog_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>level</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get hog members with respect to a given taxonomic level.

The method will return a list of protein entries that are all
member of the same hog with respect to the taxonomic range
of interest.

:param bytes hog_id: the query hog id
:param str level: the taxonomic level of interest</pre> 
</div>
</div>
<a id="af245ce87a9323b696b56ef3ca8d69811" name="af245ce87a9323b696b56ef3ca8d69811"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af245ce87a9323b696b56ef3ca8d69811">&#9670;&nbsp;</a></span>iter_members_of_hog_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.iter_members_of_hog_id </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>hog_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">iterates over all proteins that belong to a specific hog_id.

A hog_id might be an ID of the following form: HOG:0000212.1a
This method will yield all proteins in the form of
:class:`ProteinEntry` instances that are part of this hog_id.

:param str hog_id: the requested HOG ID.
:return: :py:class:`ProteinEntry` objects
:rtype: iter(:class:`ProteinEntry`)</pre> 
</div>
</div>
<a id="a7709c6ca40041c9a858153cb28788e1c" name="a7709c6ca40041c9a858153cb28788e1c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7709c6ca40041c9a858153cb28788e1c">&#9670;&nbsp;</a></span>load_gene_ontology()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.load_gene_ontology </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>factory</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>rels</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Instantiate GeneOntology object

By default, a GeneOntology object is returned based on
the default relations (which are defined in :mod:`.gene_ontology`)

The factory parameter allows to specify an subtype of
GeneOntology, e.g. :class:`.gene_ontology.FreqAwareGeneOntology`,

The rels parameter should be a list of relation strings that
should be used as parents relations.

:param factory: GeneOntology factory
:param rels: list of rels for parent relations

:returns: GeneOntology object</pre> 
</div>
</div>
<a id="a17e6f23f51a65c802055778f8aa3abdf" name="a17e6f23f51a65c802055778f8aa3abdf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a17e6f23f51a65c802055778f8aa3abdf">&#9670;&nbsp;</a></span>main_isoforms()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.main_isoforms </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>genome</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns the proteins that are the main isoforms of a genome.

The main isoform is in this context the isoform that we used in OMA to
infer the orthologs. It is the one variant that has the most alignment
matches to all other gnomes.

The genome parameter should be the UniProtSpeciesCode of the species of
interest. If it is a numeric value, the genome parameter is interpreted
as the protein entrynr. The method returns then the main isoforms for
the species to which this protein belongs.

:Note: OMA only predicts orthologs for the main isoform, so there is no
    difference if you work with only the main isoforms or all proteins of
    a genome in terms of orthologs.

:param genome: UniProtSpeciesCode of the genome of interest, or a gene
               number (EntryNr) from the genome of interest.
</pre> 
</div>
</div>
<a id="abafa5fc4a4b8d2f8df8f787d5119deca" name="abafa5fc4a4b8d2f8df8f787d5119deca"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abafa5fc4a4b8d2f8df8f787d5119deca">&#9670;&nbsp;</a></span>member_of_fam()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.member_of_fam </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>fam</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">returns an array of protein entries which belong to a given fam</pre> 
</div>
</div>
<a id="a191627558268ec8561af49ae1f1ff975" name="a191627558268ec8561af49ae1f1ff975"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a191627558268ec8561af49ae1f1ff975">&#9670;&nbsp;</a></span>member_of_hog_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.member_of_hog_id </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>hog_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>level</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">return an array of protein entries which belong to a given hog_id.

E.g. if hog_id = 'HOG122.1a', the method returns all the proteins that
have either exactly this hog id or an inparalogous id such a HOG122.1a.4b.2a

If you are only interested in the members of a specific lineage (identified
through its taxonomic range), you can pass the taxonomic range as an
additional argument. Only the proteins of genomes belonging to this clade
will be returned. Otherwise, all proteins with having this specific hog_id
will be returned.

:param str hog_id: the requested hog_id.
:param level: the taxonomic level of interest
:type level: str or None

:return: a numpy.array with the protein entries belonging to the requested hog.
:rtype: :class:`numpy.ndarray`

:Note: Even if you obtained a certain hog_id using
       :py:meth:`get_subhogids_at_level`
       using a certain level, if you do not specify the level in
       :meth:`member_of_hog_id` again, you will likely get proteins from other
       clades. Only if it happens that the deepest level of the hog_id
       coincides with the taxonomic range of interest, the two will be identical.
</pre> 
</div>
</div>
<a id="ac2688c4b6c269f99fc8974fb97ac47ab" name="ac2688c4b6c269f99fc8974fb97ac47ab"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac2688c4b6c269f99fc8974fb97ac47ab">&#9670;&nbsp;</a></span>neighbour_genes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.neighbour_genes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>entry_nr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>window</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Returns neighbor genes around a query gene.

This method returns a tuple containing a numpy recarray with
gene entries located around the query gene, and an index
pointing to the query gene. The genes are sorted according to
their position on the chromosome.

The *windows* parameter specifies the number of genes up- and
downstream of the query gene that should be reported. Note
that the actual number can be smaller if the query gene is close
to a chromosome start or end.

:param entry_nr: the entry number of the query gene
:param window: the number of neighboring genes on each
                   side to return</pre> 
</div>
</div>
<a id="a40877a4e516d4d1fdce2e1582e658b2e" name="a40877a4e516d4d1fdce2e1582e658b2e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a40877a4e516d4d1fdce2e1582e658b2e">&#9670;&nbsp;</a></span>oma_group_members()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.oma_group_members </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get the member entries of an oma group.

This method returns a numpy array of protein entries that form
an oma group. If the group id is invalid (not positive
integer value or a valid Fingerprint), an `InvalidId` Exception
is raised.

:param group_id: numeric oma group id or Fingerprint</pre> 
</div>
</div>
<a id="a4867adaf5cca9f18371f1b985097cc22" name="a4867adaf5cca9f18371f1b985097cc22"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4867adaf5cca9f18371f1b985097cc22">&#9670;&nbsp;</a></span>oma_group_metadata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def pyoma.browser.db.Database.oma_group_metadata </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>group_nr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">get the meta data associated with a OMA Group

The meta data contains the fingerprint and the keywords infered for this group.
The method retuns this information as a dictionary. The parameter must be
the numeric oma group nr.

:param int group_nr: a numeric oma group id.</pre> 
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>src/build/lib/pyoma/browser/db.py</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.4
</small></address>
</body>
</html>
