\hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter}{}\doxysection{pyoma.\+browser.\+convert.\+Darwin\+Exporter Class Reference}
\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter}\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
Inheritance diagram for pyoma.\+browser.\+convert.\+Darwin\+Exporter\+:\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[height=3.000000cm]{classpyoma_1_1browser_1_1convert_1_1DarwinExporter}
\end{center}
\end{figure}
\doxysubsection*{Public Member Functions}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aa9ad7990105c8d8d5554012e793878a1}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aa9ad7990105c8d8d5554012e793878a1}} 
def {\bfseries \+\_\+\+\_\+init\+\_\+\+\_\+} (self, path, logger=None, mode=None)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a176d39fb4170f8ce6b46bc7485901032}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a176d39fb4170f8ce6b46bc7485901032}} 
def {\bfseries call\+\_\+darwin\+\_\+export} (self, func)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_af51693db81563fb7cf414508d169bd64}{create\+\_\+table\+\_\+if\+\_\+needed}} (self, parent, name, drop\+\_\+data=False, $\ast$$\ast$kwargs)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a64d076922eb37d7cd93311ebcc78b576}{get\+\_\+version}} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a6d7aeff94950cf6274b488cb8341dc8f}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a6d7aeff94950cf6274b488cb8341dc8f}} 
def {\bfseries add\+\_\+version} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a9ebadc86fe750f3f1f95d3a469cd0b76}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a9ebadc86fe750f3f1f95d3a469cd0b76}} 
def {\bfseries add\+\_\+species\+\_\+data} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ad3734c567afa6ddf3ad4a5f9c34d8009}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ad3734c567afa6ddf3ad4a5f9c34d8009}} 
def {\bfseries add\+\_\+orthologs} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ad3013fb2193e05b492c2c96c3e81eb5f}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ad3013fb2193e05b492c2c96c3e81eb5f}} 
def {\bfseries add\+\_\+same\+\_\+species\+\_\+relations} (self)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ae41d79319884a524458096ec4ad8c78f}{add\+\_\+synteny\+\_\+scores}} (self)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a526b5627a743b04104219500060a871c}{add\+\_\+homoeology\+\_\+confidence}} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a041a84b7efb5ce6df4d84711a7a99acb}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a041a84b7efb5ce6df4d84711a7a99acb}} 
def {\bfseries add\+\_\+proteins} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a991d2a0f4d7be6aa74b4674ea21a11e5}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a991d2a0f4d7be6aa74b4674ea21a11e5}} 
def {\bfseries add\+\_\+hogs} (self)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a09ee4b5dbb8b4abd7da7bd32010190e8}{add\+\_\+orthoxml}} (self, orthoxml\+\_\+path, fam\+\_\+nrs)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a6a6f1cfdc09fd18ff0b5547cdf2794d3}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a6a6f1cfdc09fd18ff0b5547cdf2794d3}} 
def {\bfseries xref\+\_\+databases} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a7f80d423709fa9f92c18d44e04a2e449}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a7f80d423709fa9f92c18d44e04a2e449}} 
def {\bfseries add\+\_\+xrefs} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a821692e070579346ebcdf44afe67b29b}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a821692e070579346ebcdf44afe67b29b}} 
def {\bfseries add\+\_\+group\+\_\+metadata} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aa013135f187257ff446fca7407ea22e3}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aa013135f187257ff446fca7407ea22e3}} 
def {\bfseries close} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_acf1f5b026f51925ff13e6ae775ce4397}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_acf1f5b026f51925ff13e6ae775ce4397}} 
def {\bfseries create\+\_\+indexes} (self)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a578fb011d7ceaa073b26aedc91101c07}{add\+\_\+canonical\+\_\+id}} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ab0d59c3153a68d3dcd937cc9838224c8}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ab0d59c3153a68d3dcd937cc9838224c8}} 
def {\bfseries add\+\_\+domain\+\_\+info} (self, domains)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a41d66ee6970ebe2b4213968d0cee2727}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a41d66ee6970ebe2b4213968d0cee2727}} 
def {\bfseries add\+\_\+domainname\+\_\+info} (self, domainname\+\_\+infos)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a6c3790ecf5cc573ef300c86e35aee92b}{update\+\_\+summary\+\_\+stats}} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a9b482b7a1f534c328d4e7ea2099a7d8c}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a9b482b7a1f534c328d4e7ea2099a7d8c}} 
def {\bfseries count\+\_\+gene\+\_\+ontology\+\_\+summary} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a0385ea373dab915019b8fc218b5de8ba}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a0385ea373dab915019b8fc218b5de8ba}} 
def {\bfseries count\+\_\+xref\+\_\+summary} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a0008bfc32526197faad34779aa1d1849}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a0008bfc32526197faad34779aa1d1849}} 
def {\bfseries collect\+\_\+group\+\_\+sizes} (self)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a8d4a1ed415bf09caebc594e5b294e870}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a8d4a1ed415bf09caebc594e5b294e870}} 
def {\bfseries compute\+\_\+domaincovered\+\_\+sites} (self)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a92b0b44e8bab080c865cd3495f550229}{add\+\_\+domain\+\_\+covered\+\_\+sites\+\_\+counts}} (self)
\item 
def \mbox{\hyperlink{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a78dabb8c29af0ad6407f9f42fce8fb54}{add\+\_\+sequence\+\_\+suffix\+\_\+array}} (self, k=6, fn=None, sa=None)
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a0eefa340bebb62c6856c0ea5fac29567}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a0eefa340bebb62c6856c0ea5fac29567}} 
def {\bfseries add\+\_\+hog\+\_\+domain\+\_\+prevalence} (self)
\end{DoxyCompactItemize}
\doxysubsection*{Public Attributes}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a48e0f3bad4834ca08ea532d5333cbaef}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a48e0f3bad4834ca08ea532d5333cbaef}} 
{\bfseries logger}
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a20cbfaf2ad0cd2e13a0e70cde720b84d}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a20cbfaf2ad0cd2e13a0e70cde720b84d}} 
{\bfseries h5}
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ac3a77e2a4ce0a01f0a3eed4b03a1b375}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ac3a77e2a4ce0a01f0a3eed4b03a1b375}} 
{\bfseries orthoxml\+\_\+buffer}
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aaa226a121f49989aed2121b60af016d5}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aaa226a121f49989aed2121b60af016d5}} 
{\bfseries orthoxml\+\_\+index}
\end{DoxyCompactItemize}
\doxysubsection*{Static Public Attributes}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a8562fffc3d9a9864389d14428ff7a753}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a8562fffc3d9a9864389d14428ff7a753}} 
string {\bfseries DB\+\_\+\+SCHEMA\+\_\+\+VERSION} = \textquotesingle{}3.\+2\textquotesingle{}
\item 
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aada2e9f337a9abc1a83fbd4539ddb7b9}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_aada2e9f337a9abc1a83fbd4539ddb7b9}} 
{\bfseries DRW\+\_\+\+CONVERT\+\_\+\+FILE} = os.\+path.\+abspath(os.\+path.\+splitext(\+\_\+\+\_\+file\+\_\+\+\_\+)\mbox{[}0\mbox{]} + \textquotesingle{}.drw\textquotesingle{})
\end{DoxyCompactItemize}


\doxysubsection{Member Function Documentation}
\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a578fb011d7ceaa073b26aedc91101c07}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a578fb011d7ceaa073b26aedc91101c07}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!add\_canonical\_id@{add\_canonical\_id}}
\index{add\_canonical\_id@{add\_canonical\_id}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{add\_canonical\_id()}{add\_canonical\_id()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+add\+\_\+canonical\+\_\+id (\begin{DoxyParamCaption}\item[{}]{self }\end{DoxyParamCaption})}

\begin{DoxyVerb}add one canonical xref id to the /Protein/Entries table.\end{DoxyVerb}
 \mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a92b0b44e8bab080c865cd3495f550229}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a92b0b44e8bab080c865cd3495f550229}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!add\_domain\_covered\_sites\_counts@{add\_domain\_covered\_sites\_counts}}
\index{add\_domain\_covered\_sites\_counts@{add\_domain\_covered\_sites\_counts}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{add\_domain\_covered\_sites\_counts()}{add\_domain\_covered\_sites\_counts()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+add\+\_\+domain\+\_\+covered\+\_\+sites\+\_\+counts (\begin{DoxyParamCaption}\item[{}]{self }\end{DoxyParamCaption})}

\begin{DoxyVerb}Stores the number of AA covered by a DomainAnnotation.

This method adds to the hdf5 file a /Protein/DomainCoverage array that
contains the number of AA sites covered by a domain. The position
corresponds to the protein entry numbers in /Protein/Entries.

:Note: The method assumes that the domains are all non-overlapping.
    If they are not, the reported coverage will be too high!

:return: covered fractions by domains for each protein
:rtype: numpy.array\end{DoxyVerb}
 \mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a526b5627a743b04104219500060a871c}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a526b5627a743b04104219500060a871c}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!add\_homoeology\_confidence@{add\_homoeology\_confidence}}
\index{add\_homoeology\_confidence@{add\_homoeology\_confidence}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{add\_homoeology\_confidence()}{add\_homoeology\_confidence()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+add\+\_\+homoeology\+\_\+confidence (\begin{DoxyParamCaption}\item[{}]{self }\end{DoxyParamCaption})}

\begin{DoxyVerb}adds the homoeology confidence scores to the database.

This method should be called only after the synteny scores have
been computed and added to the database.

The computations are done using :mod:`homoeologs` module.\end{DoxyVerb}
 \mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a09ee4b5dbb8b4abd7da7bd32010190e8}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a09ee4b5dbb8b4abd7da7bd32010190e8}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!add\_orthoxml@{add\_orthoxml}}
\index{add\_orthoxml@{add\_orthoxml}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{add\_orthoxml()}{add\_orthoxml()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+add\+\_\+orthoxml (\begin{DoxyParamCaption}\item[{}]{self,  }\item[{}]{orthoxml\+\_\+path,  }\item[{}]{fam\+\_\+nrs }\end{DoxyParamCaption})}

\begin{DoxyVerb}append orthoxml file content to orthoxml_buffer array and add index for the HOG family\end{DoxyVerb}
 \mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a78dabb8c29af0ad6407f9f42fce8fb54}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a78dabb8c29af0ad6407f9f42fce8fb54}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!add\_sequence\_suffix\_array@{add\_sequence\_suffix\_array}}
\index{add\_sequence\_suffix\_array@{add\_sequence\_suffix\_array}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{add\_sequence\_suffix\_array()}{add\_sequence\_suffix\_array()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+add\+\_\+sequence\+\_\+suffix\+\_\+array (\begin{DoxyParamCaption}\item[{}]{self,  }\item[{}]{k = {\ttfamily 6},  }\item[{}]{fn = {\ttfamily None},  }\item[{}]{sa = {\ttfamily None} }\end{DoxyParamCaption})}

\begin{DoxyVerb}    Adds the sequence suffix array to the database. NOTE: this
    (obviously) requires A LOT of memory for large DBs.
\end{DoxyVerb}
 \mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ae41d79319884a524458096ec4ad8c78f}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_ae41d79319884a524458096ec4ad8c78f}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!add\_synteny\_scores@{add\_synteny\_scores}}
\index{add\_synteny\_scores@{add\_synteny\_scores}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{add\_synteny\_scores()}{add\_synteny\_scores()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+add\+\_\+synteny\+\_\+scores (\begin{DoxyParamCaption}\item[{}]{self }\end{DoxyParamCaption})}

\begin{DoxyVerb}add synteny scores of pairwise relations to database.

Current implementation only computes synteny scores for
homoeologs, but easy to extend. Question is rather if we
need synteny scores for all genome pairs, and if not, how
to select.

The computations of the scores are done using :mod:`synteny`
module of this package.\end{DoxyVerb}
 \mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_af51693db81563fb7cf414508d169bd64}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_af51693db81563fb7cf414508d169bd64}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!create\_table\_if\_needed@{create\_table\_if\_needed}}
\index{create\_table\_if\_needed@{create\_table\_if\_needed}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{create\_table\_if\_needed()}{create\_table\_if\_needed()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+create\+\_\+table\+\_\+if\+\_\+needed (\begin{DoxyParamCaption}\item[{}]{self,  }\item[{}]{parent,  }\item[{}]{name,  }\item[{}]{drop\+\_\+data = {\ttfamily False},  }\item[{$\ast$$\ast$}]{kwargs }\end{DoxyParamCaption})}

\begin{DoxyVerb}create a table if needed.

The function only checks whether a table exists with that name,
but not if it is compatible with the passed arguments.
if you pass data with the `obj` argument, this data is appended
to the table. If you set `drop_data` to True, data that was
previously in the existing table is dropped prior to adding new
data.\end{DoxyVerb}
 \mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a64d076922eb37d7cd93311ebcc78b576}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a64d076922eb37d7cd93311ebcc78b576}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!get\_version@{get\_version}}
\index{get\_version@{get\_version}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{get\_version()}{get\_version()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+get\+\_\+version (\begin{DoxyParamCaption}\item[{}]{self }\end{DoxyParamCaption})}

\begin{DoxyVerb}return version of the dataset.

Default implementation searches for 'mname' in Matrix or matrix_stats.drw files.
\end{DoxyVerb}
 

Reimplemented in \mbox{\hyperlink{classpyoma_1_1browser_1_1convert__omastandalone_1_1StandaloneExporter_a5afadc1f24975add301a188afc23ad43}{pyoma.\+browser.\+convert\+\_\+omastandalone.\+Standalone\+Exporter}}.

\mbox{\Hypertarget{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a6c3790ecf5cc573ef300c86e35aee92b}\label{classpyoma_1_1browser_1_1convert_1_1DarwinExporter_a6c3790ecf5cc573ef300c86e35aee92b}} 
\index{pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}!update\_summary\_stats@{update\_summary\_stats}}
\index{update\_summary\_stats@{update\_summary\_stats}!pyoma.browser.convert.DarwinExporter@{pyoma.browser.convert.DarwinExporter}}
\doxysubsubsection{\texorpdfstring{update\_summary\_stats()}{update\_summary\_stats()}}
{\footnotesize\ttfamily def pyoma.\+browser.\+convert.\+Darwin\+Exporter.\+update\+\_\+summary\+\_\+stats (\begin{DoxyParamCaption}\item[{}]{self }\end{DoxyParamCaption})}

\begin{DoxyVerb}update the summary statistics of xrefs & go.

The function analyses the well-known xref sources as well as
GO annotations and computes aggregated counts for
all / in OMA Group / in HOGs for all of them.
\end{DoxyVerb}
 

The documentation for this class was generated from the following file\+:\begin{DoxyCompactItemize}
\item 
src/build/lib/pyoma/browser/convert.\+py\end{DoxyCompactItemize}
