  % Handle batch mode
  % to overcome problems with too many open files
  \let\mypdfximage\pdfximage\def\pdfximage{\immediate\mypdfximage}
  % Set document class depending on configuration
  \documentclass[twoside]{book}
  %% moved from doxygen.sty due to workaround for LaTex 2019 version and unmaintained tabu package
  \usepackage{ifthen}
  \ifx\requestedLaTeXdate\undefined
    \usepackage{array}
  \else
    \usepackage{array}[=2016-10-06]
  \fi
  %%
  % Packages required by doxygen
  \usepackage{fixltx2e} % for \textsubscript
  \usepackage{doxygen}
  \usepackage{graphicx}
  \usepackage[utf8]{inputenc}
  \usepackage{makeidx}
  \PassOptionsToPackage{warn}{textcomp}
  \usepackage{textcomp}
  \usepackage[nointegrals]{wasysym}
  \usepackage{ifxetex}
  % NLS support packages
  % Define default fonts
  % Font selection
  \usepackage[T1]{fontenc}
  % set main and monospaced font
  \usepackage[scaled=.90]{helvet}
\usepackage{courier}
\renewcommand{\familydefault}{\sfdefault}
  \usepackage{sectsty}
  \allsectionsfont{%
    \fontseries{bc}\selectfont%
    \color{darkgray}%
  }
  \renewcommand{\DoxyLabelFont}{%
    \fontseries{bc}\selectfont%
    \color{darkgray}%
  }
  \newcommand{\+}{\discretionary{\mbox{\scriptsize$\hookleftarrow$}}{}{}}
   % Arguments of doxygenemoji:
   % 1) ':<text>:' form of the emoji, already LaTeX-escaped
   % 2) file with the name of the emoji without the .png extension
   % in case image exist use this otherwise use the ':<text>:' form
   \newcommand{\doxygenemoji}[2]{%
     \IfFileExists{./#2.png}{\raisebox{-0.1em}{\includegraphics[height=0.9em]{./#2.png}}}{#1}%
   }
  % Page & text layout
  \usepackage{geometry}
  \geometry{%
    a4paper,%
    top=2.5cm,%
    bottom=2.5cm,%
    left=2.5cm,%
    right=2.5cm%
  }
  % Allow a bit of overflow to go unnoticed by other means
  \tolerance=750
  \hfuzz=15pt
  \hbadness=750
  \setlength{\emergencystretch}{15pt}
  \setlength{\parindent}{0cm}
  \newcommand{\doxynormalparskip}{\setlength{\parskip}{3ex plus 2ex minus 2ex}}
  \newcommand{\doxytocparskip}{\setlength{\parskip}{1ex plus 0ex minus 0ex}}
  \doxynormalparskip
  % Redefine paragraph/subparagraph environments, using sectsty fonts
  \makeatletter
  \renewcommand{\paragraph}{%
    \@startsection{paragraph}{4}{0ex}{-1.0ex}{1.0ex}{%
      \normalfont\normalsize\bfseries\SS@parafont%
    }%
  }
  \renewcommand{\subparagraph}{%
    \@startsection{subparagraph}{5}{0ex}{-1.0ex}{1.0ex}{%
      \normalfont\normalsize\bfseries\SS@subparafont%
    }%
  }
  \makeatother
  \makeatletter
  \newcommand\hrulefilll{\leavevmode\leaders\hrule\hskip 0pt plus 1filll\kern\z@}
  \makeatother
  % Headers & footers
  \usepackage{fancyhdr}
  \pagestyle{fancyplain}
  \renewcommand{\footrulewidth}{0.4pt}
  \fancypagestyle{fancyplain}{
    \fancyhf{}
    \fancyhead[LE, RO]{\bfseries\thepage}
    \fancyhead[LO]{\bfseries\rightmark}
    \fancyhead[RE]{\bfseries\leftmark}
    \fancyfoot[LO, RE]{\bfseries\scriptsize Generated by Doxygen }
  }
  \fancypagestyle{plain}{
    \fancyhf{}
    \fancyfoot[LO, RE]{\bfseries\scriptsize Generated by Doxygen }
    \renewcommand{\headrulewidth}{0pt}
  }
  \pagestyle{fancyplain}
  \renewcommand{\chaptermark}[1]{%
    \markboth{#1}{}%
  }
  \renewcommand{\sectionmark}[1]{%
    \markright{\thesection\ #1}%
  }
  % ToC, LoF, LoT, bibliography, and index
  % Indices & bibliography
  \usepackage{natbib}
  \usepackage[titles]{tocloft}
  \setcounter{tocdepth}{3}
  \setcounter{secnumdepth}{5}
  % creating indexes
  \makeindex
  \usepackage{newunicodechar}
  \newunicodechar{⁻}{${}^{-}$}% Superscript minus
  \newunicodechar{²}{${}^{2}$}% Superscript two
  \newunicodechar{³}{${}^{3}$}% Superscript three
  % Hyperlinks
    % Hyperlinks (required, but should be loaded last)
    \ifpdf
      \usepackage[pdftex,pagebackref=true]{hyperref}
    \else
      \ifxetex
        \usepackage[pagebackref=true]{hyperref}
      \else
        \usepackage[ps2pdf,pagebackref=true]{hyperref}
      \fi
    \fi
    \hypersetup{%
      colorlinks=true,%
      linkcolor=blue,%
      citecolor=blue,%
      unicode,%
      pdftitle={Hog\+Prof},%
      pdfsubject={}%
    }
  % Custom commands used by the header
  % Custom commands
  \newcommand{\clearemptydoublepage}{%
    \newpage{\pagestyle{empty}\cleardoublepage}%
  }
  % caption style definition
  \usepackage{caption}
  \captionsetup{labelsep=space,justification=centering,font={bf},singlelinecheck=off,skip=4pt,position=top}
  % in page table of contents
  \usepackage{etoc}
  \etocsettocstyle{\doxytocparskip}{\doxynormalparskip}
  % prevent numbers overlap the titles in toc
  \renewcommand{\numberline}[1]{#1~}
% End of preamble, now comes the document contents
%===== C O N T E N T S =====
\begin{document}
  \raggedbottom
  % Titlepage & ToC
    % To avoid duplicate page anchors due to reuse of same numbers for
    % the index (be it as roman numbers)
    \hypersetup{pageanchor=false,
                bookmarksnumbered=true,
                pdfencoding=unicode
               }
  \pagenumbering{alph}
  \begin{titlepage}
  \vspace*{7cm}
  \begin{center}%
  {\Large Hog\+Prof}\\
  \vspace*{1cm}
  {\large Generated by Doxygen 1.9.4}\\
  \end{center}
  \end{titlepage}
  \clearemptydoublepage
  \pagenumbering{roman}
  \tableofcontents
  \clearemptydoublepage
  \pagenumbering{arabic}
  % re-enable anchors again
  \hypersetup{pageanchor=true}
%--- Begin generated contents ---
\chapter{Namespace Index}
\input{namespaces}
\chapter{Hierarchical Index}
\input{hierarchy}
\chapter{Class Index}
\input{annotated}
\chapter{Namespace Documentation}
\input{namespacepyoma_1_1browser_1_1KmerEncoder}
\chapter{Class Documentation}
\input{classpyoma_1_1browser_1_1geneontology_1_1AbstractParser}
\input{classpyoma_1_1browser_1_1db_1_1AmbiguousID}
\input{classpyoma_1_1browser_1_1geneontology_1_1AnnotationFilter}
\input{classpyoma_1_1browser_1_1geneontology_1_1AnnotationParser}
\input{classpyoma_1_1browser_1_1convert_1_1CathDomainNameParser}
\input{classpyoma_1_1browser_1_1convert_1_1DarwinDbEntryParser}
\input{classpyoma_1_1browser_1_1convert_1_1DarwinException}
\input{classpyoma_1_1browser_1_1convert_1_1DarwinExporter}
\input{classpyoma_1_1browser_1_1db_1_1Database}
\input{classsrc_1_1orthoxml_1_1database}
\input{classpyoma_1_1browser_1_1check__db__consistency_1_1DatabaseChecks}
\input{classpyoma_1_1browser_1_1convert_1_1DataImportError}
\input{classpyoma_1_1browser_1_1db_1_1DBConsistencyError}
\input{classpyoma_1_1browser_1_1db_1_1DBVersionError}
\input{classpyoma_1_1browser_1_1convert_1_1DescriptionManager}
\input{classpyoma_1_1browser_1_1tablefmt_1_1DomainDescriptionTable}
\input{classpyoma_1_1browser_1_1db_1_1DomainNameIdMapper}
\input{classpyoma_1_1browser_1_1tablefmt_1_1DomainTable}
\input{classpyoma_1_1browser_1_1tablefmt_1_1ECTable}
\input{classpyoma_1_1browser_1_1models_1_1Exon}
\input{classpyoma_1_1browser_1_1models_1_1ExonStructure}
\input{classpyoma_1_1browser_1_1db_1_1FastMapper}
\input{classpyoma_1_1browser_1_1geneontology_1_1FreqAwareGeneOntology}
\input{classsrc_1_1orthoxml_1_1GDSParseError}
\input{classsrc_1_1orthoxml_1_1gene}
\input{classpyoma_1_1browser_1_1geneontology_1_1GeneOntology}
\input{classpyoma_1_1browser_1_1models_1_1GeneOntologyAnnotation}
\input{classpyoma_1_1browser_1_1convert_1_1GeneOntologyManager}
\input{classpyoma_1_1browser_1_1tablefmt_1_1GeneOntologyTable}
\input{classsrc_1_1orthoxml_1_1GeneratedsSuper}
\input{classsrc_1_1orthoxml_1_1geneRef}
\input{classsrc_1_1orthoxml_1_1genes}
\input{classpyoma_1_1browser_1_1linkout_1_1GenesPriorizationHandler}
\input{classpyoma_1_1browser_1_1linkout_1_1GenesResource}
\input{classpyoma_1_1browser_1_1models_1_1Genome}
\input{classpyoma_1_1browser_1_1tablefmt_1_1GenomeTable}
\input{classpyoma_1_1browser_1_1geneontology_1_1GOAspect}
\input{classpyoma_1_1browser_1_1geneontology_1_1GOterm}
\input{classsrc_1_1orthoxml_1_1group}
\input{classpyoma_1_1browser_1_1convert_1_1GroupAnnotatorInclGeneRefs}
\input{classsrc_1_1orthoxml_1_1groups}
\input{classpyoma_1_1browser_1_1tablefmt_1_1GroupsizeHistogram}
\input{classsrc_1_1tests_1_1test__hashutils_1_1HashutilsTest}
\input{classtests_1_1test__hashutils_1_1HashutilsTest}
\input{classpyoma_1_1browser_1_1convert_1_1HogConverter}
\input{classpyoma_1_1browser_1_1tablefmt_1_1HOGDomainArchPrevalenceTable}
\input{classpyoma_1_1browser_1_1tablefmt_1_1HOGDomainPresenceTable}
\input{classpyoma_1_1browser_1_1tablefmt_1_1HOGsTable}
\input{classpyoma_1_1browser_1_1homoeologs_1_1HomeologsConfidenceCalculator}
\input{classpyoma_1_1browser_1_1homoeologs_1_1HomeologsConfidenceCalculatorFromTSV}
\input{classsrc_1_1tests_1_1test__hpputils_1_1HpputilsTest}
\input{classtests_1_1test__hpputils_1_1HpputilsTest}
\input{classpyoma_1_1browser_1_1db_1_1IdMapperFactory}
\input{classpyoma_1_1browser_1_1db_1_1IDResolver}
\input{classpyoma_1_1browser_1_1db_1_1InvalidId}
\input{classpyoma_1_1browser_1_1db_1_1InvalidOmaId}
\input{classpyoma_1_1browser_1_1db_1_1InvalidTaxonId}
\input{classpyoma_1_1browser_1_1models_1_1KeyWrapper}
\input{classpyoma_1_1browser_1_1KmerEncoder_1_1KmerEncoder}
\input{classpyoma_1_1browser_1_1models_1_1LazyProperty}
\input{classpyoma_1_1browser_1_1linkout_1_1LinkoutBuffer}
\input{classpyoma_1_1browser_1_1db_1_1LinkoutIdMapper}
\input{classpyoma_1_1browser_1_1locus__parser_1_1LocusParser}
\input{classpyoma_1_1browser_1_1tablefmt_1_1LocusTable}
\input{classpyoma_1_1browser_1_1locus__parser_1_1LocusTransformer}
\input{classsrc_1_1lshbuilder_1_1LSHBuilder}
\input{classsrc_1_1orthoxml_1_1MemberSpec__}
\input{classsrc_1_1orthoxml_1_1MixedContainer}
\input{classpyoma_1_1browser_1_1linkout_1_1NCBILinkOutXML}
\input{classpyoma_1_1browser_1_1db_1_1NoReprEntry}
\input{classsrc_1_1orthoxml_1_1notes}
\input{classpyoma_1_1browser_1_1convert_1_1OmaGroupMetadataLoader}
\input{classpyoma_1_1browser_1_1tablefmt_1_1OmaGroupTable}
\input{classpyoma_1_1browser_1_1db_1_1OmaIdMapper}
\input{classpyoma_1_1browser_1_1geneontology_1_1OntologyParser}
\input{classsrc_1_1orthoxml_1_1orthoXML}
\input{classpyoma_1_1browser_1_1tablefmt_1_1OrthoXmlHogTable}
\input{classpyoma_1_1browser_1_1OrthoXMLSplitter_1_1OrthoXMLSplitter}
\input{classpyoma_1_1browser_1_1models_1_1PairwiseRelation}
\input{classpyoma_1_1browser_1_1tablefmt_1_1PairwiseRelationTable}
\input{classpyoma_1_1browser_1_1convert_1_1PfamDomainNameParser}
\input{classsrc_1_1profiler_1_1Profiler}
\input{classsrc_1_1tests_1_1test__profiler_1_1ProfilerTest}
\input{classtests_1_1test__profiler_1_1ProfilerTest}
\input{classsrc_1_1orthoxml_1_1property}
\input{classpyoma_1_1browser_1_1models_1_1ProteinEntry}
\input{classpyoma_1_1browser_1_1linkout_1_1ProteinResource}
\input{classpyoma_1_1browser_1_1tablefmt_1_1ProteinTable}
\input{classpyoma_1_1browser_1_1linkout_1_1Provider}
\input{classpyoma_1_1browser_1_1linkout_1_1Resource}
\input{classsrc_1_1orthoxml_1_1score}
\input{classsrc_1_1orthoxml_1_1scoreDef}
\input{classsrc_1_1orthoxml_1_1scores}
\input{classpyoma_1_1browser_1_1db_1_1SequenceSearch}
\input{classgo2vec__train_1_1SimilarityCallback}
\input{classprot2vec_1_1SimilarityCallback}
\input{classpyoma_1_1browser_1_1db_1_1Singleton}
\input{classpyoma_1_1browser_1_1models_1_1Singleton}
\input{classsrc_1_1orthoxml_1_1species}
\input{classpyoma_1_1browser_1_1convert__omastandalone_1_1StandaloneExporter}
\input{classpyoma_1_1browser_1_1db_1_1SuffixSearcher}
\input{classpyoma_1_1browser_1_1synteny_1_1SyntenyScorer}
\input{classpyoma_1_1browser_1_1db_1_1Taxonomy}
\input{classpyoma_1_1browser_1_1linkout_1_1TaxonomyResource}
\input{classpyoma_1_1browser_1_1tablefmt_1_1TaxonomyTable}
\input{classpyoma_1_1browser_1_1synteny_1_1TooSmallChromosome}
\input{classpyoma_1_1browser_1_1db_1_1UniProtIdMapper}
\input{classpyoma_1_1browser_1_1db_1_1UnknownIdType}
\input{classpyoma_1_1browser_1_1db_1_1UnknownSpecies}
\input{classsrc_1_1validation_1_1validation__semantic__similarity_1_1Validation__semantic__similarity}
\input{classvalidation_1_1validation__semantic__similarity_1_1Validation__semantic__similarity}
\input{classpyoma_1_1browser_1_1db_1_1XrefIdMapper}
\input{classpyoma_1_1browser_1_1convert_1_1XRefImporter}
\input{classpyoma_1_1browser_1_1tablefmt_1_1XRefTable}
%--- End generated contents ---
% Index
  \backmatter
  \newpage
  \phantomsection
  \clearemptydoublepage
  \addcontentsline{toc}{chapter}{\indexname}
  \printindex
% Required for some languages (in combination with latexdocumentpre from the header)
\end{document}
