
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml" lang="Python">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Welcome to Getting Started with Sphinx&#39;s documentation! &#8212; PyProfiler alpha documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-pyprofiler.kerasopt">
<span id="welcome-to-getting-started-with-sphinx-s-documentation"></span><h1>Welcome to Getting Started with Sphinx's documentation!<a class="headerlink" href="#module-pyprofiler.kerasopt" title="Permalink to this headline">¶</a></h1>
<span class="target" id="module-pyprofiler.lshbuilder"></span><dl class="class">
<dt id="pyprofiler.lshbuilder.LSHBuilder">
<em class="property">class </em><code class="descclassname">pyprofiler.lshbuilder.</code><code class="descname">LSHBuilder</code><span class="sig-paren">(</span><em>tarfile_ortho=None</em>, <em>h5_oma=None</em>, <em>taxa=None</em>, <em>masterTree=None</em>, <em>saving_name=None</em>, <em>numperm=256</em>, <em>treeweights=None</em>, <em>taxfilter=None</em>, <em>taxmask=None</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.lshbuilder.LSHBuilder" title="Permalink to this definition">¶</a></dt>
<dd><p>This class contains the stuff you need to make a phylogenetic profiling database with input orthxml files and a taxonomic tree</p>
<p>You must either input an OMA hdf5 file or provide a function to generate orthoxml files for each orthologous groupself.</p>
<p>You can provide a species tree or use the ncbi taxonomy with a list of taxonomic codes for all the species in your db</p>
</dd></dl>

<span class="target" id="module-pyprofiler.profiler"></span><dl class="class">
<dt id="pyprofiler.profiler.Profiler">
<em class="property">class </em><code class="descclassname">pyprofiler.profiler.</code><code class="descname">Profiler</code><span class="sig-paren">(</span><em>lshforestpath=None</em>, <em>hashes_h5=None</em>, <em>mat_path=None</em>, <em>oma=False</em>, <em>tar=None</em>, <em>nsamples=256</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler" title="Permalink to this definition">¶</a></dt>
<dd><p>A profiler object allows the user to query the LSH with HOGs and get a list of result HOGs back</p>
<dl class="staticmethod">
<dt id="pyprofiler.profiler.Profiler.allvall_hashes">
<em class="property">static </em><code class="descname">allvall_hashes</code><span class="sig-paren">(</span><em>hashes</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.allvall_hashes" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dict of hogs:hashes, returns generate an all v all jaccard distance matrix.
:param hashes: a dict of hogs:hashes
:return: hashmat</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.allvall_nx">
<code class="descname">allvall_nx</code><span class="sig-paren">(</span><em>hashes</em>, <em>thresh=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.allvall_nx" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dict of hogs:hashes, returns generate an all v all jaccard distance matrix.
:param hashes: a dict of hogs:hashes
:return: hashmat</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.get_vpairs">
<code class="descname">get_vpairs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.get_vpairs" title="Permalink to this definition">¶</a></dt>
<dd><p>get pairwise distance matrix of OMA all v all
#not finished
:param fam: an oma fam
:return sparesemat: a mat with all taxa in Oma with nonzero entries where this protein is found
:return densemat: a mat with the taxa covered by the fam</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.hog_query">
<code class="descname">hog_query</code><span class="sig-paren">(</span><em>hog_id=None</em>, <em>fam_id=None</em>, <em>k=100</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.hog_query" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a hog_id or a fam_id as a query, returns a dictionary containing the results of the LSH.
:param hog_id: query hog id
:param fam_id: query fam id
:return: list containing the results of the LSH for the given query</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.hog_query_OMA">
<code class="descname">hog_query_OMA</code><span class="sig-paren">(</span><em>hog_id=None</em>, <em>fam_id=None</em>, <em>k=100</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.hog_query_OMA" title="Permalink to this definition">¶</a></dt>
<dd><p>Untested, Given a hog_id or a fam_id as a query, returns a dictionary containing the results of the LSH.
Generates the tree profile and hashes on the fly
:param hog_id: query hog id
:param fam_id: query fam id
:return: list containing the results of the LSH for the given query</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.hog_query_sorted">
<code class="descname">hog_query_sorted</code><span class="sig-paren">(</span><em>hog_id=None</em>, <em>fam_id=None</em>, <em>k=100</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.hog_query_sorted" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a hog_id or a fam_id as a query, returns a dictionary containing the results of the LSH.
:param hog_id: query hog id
:param fam_id: query fam id
:return: list containing the results of the LSH for the given query</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.hog_v_hog">
<code class="descname">hog_v_hog</code><span class="sig-paren">(</span><em>hogs</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.hog_v_hog" title="Permalink to this definition">¶</a></dt>
<dd><p>give two hogs returns jaccard distance.
:param hog1 , hog2: str hog id
:return: jaccard score</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.pull_hashes">
<code class="descname">pull_hashes</code><span class="sig-paren">(</span><em>hoglist</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.pull_hashes" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a list of hog_ids , returns a dictionary containing their hashes.
This uses the hdf5 file to get the hashvalues
:param hog_id: query hog id
:param fam_id: query fam id
:return: a dict containing the hash values of the hogs in hoglist</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.pull_matrows">
<code class="descname">pull_matrows</code><span class="sig-paren">(</span><em>fams</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.pull_matrows" title="Permalink to this definition">¶</a></dt>
<dd><p>given a list of fams return the submatrix containing their profiles</p>
<p>:return:fams sorted, sparse mat</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.retmat_mp">
<code class="descname">retmat_mp</code><span class="sig-paren">(</span><em>traindf</em>, <em>nworkers=25</em>, <em>chunksize=50</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.retmat_mp" title="Permalink to this definition">¶</a></dt>
<dd><p>function used to create training matrix with pairs of hogs. calculate_x will return the difference and union of
two binary vectors generated by pyham</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.retmat_mp_profiles">
<code class="descname">retmat_mp_profiles</code><span class="sig-paren">(</span><em>fams</em>, <em>nworkers=25</em>, <em>chunksize=50</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.retmat_mp_profiles" title="Permalink to this definition">¶</a></dt>
<dd><p>function used to create dataframe containing binary profiles
and trees of fams</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.return_profile_OTF">
<code class="descname">return_profile_OTF</code><span class="sig-paren">(</span><em>fam</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.return_profile_OTF" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns profiles as binary vectors for use with optimisation pipelines</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.return_profile_OTF_DCA">
<code class="descname">return_profile_OTF_DCA</code><span class="sig-paren">(</span><em>fam</em>, <em>lock=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.return_profile_OTF_DCA" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns profiles as strings for use with DCA pipelines
just concatenate the numpy arrays and use the tostring
function to generate an input &quot;alignment&quot;</p>
</dd></dl>

<dl class="staticmethod">
<dt id="pyprofiler.profiler.Profiler.sort_hashes">
<em class="property">static </em><code class="descname">sort_hashes</code><span class="sig-paren">(</span><em>query_hash</em>, <em>hashes</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.sort_hashes" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dict of hogs:hashes, returns a sorted array of hogs and jaccard distances relative to query hog.
:param query hash: weighted minhash of the query
:param hashes: a dict of hogs:hashes
:return: sortedhogs, jaccard</p>
</dd></dl>

<dl class="method">
<dt id="pyprofiler.profiler.Profiler.worker">
<code class="descname">worker</code><span class="sig-paren">(</span><em>i</em>, <em>inq</em>, <em>retq</em><span class="sig-paren">)</span><a class="headerlink" href="#pyprofiler.profiler.Profiler.worker" title="Permalink to this definition">¶</a></dt>
<dd><p>this worker function is for parallelization of generation of binary vector for use with optimisation pipelines</p>
</dd></dl>

</dd></dl>

<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">PyProfiler</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Dave Moi.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.5</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>